components/raftstore/src/store/peer.rs

```diff
@@ -704,6 +726,326 @@ pub enum UnsafeRecoveryState {
+            debug!(
+                "prevents renew lease while splitting";
+                "region_id" => self.region_id(),
+                "peer_id" => self.peer_id(),
+            );
+            None
+        } else if self.is_merging() {
+            // A merging leader should not renew its lease.
+            // Because we merge regions asynchronous, the leader may read stale results
+            // if commit merge runs slow on sibling peers.
+            debug!(
+                "prevents renew lease while merging";
+                "region_id" => self.region_id(),
+                "peer_id" => self.peer_id(),
+            );
+            None
+        } else if self.has_force_leader() {
+            debug!(
+                "prevents renew lease while in force leader state";
+                "region_id" => self.region_id(),
+                "peer_id" => self.peer_id(),
+            );
+            None
+        } else {
+            self.mut_leader_lease().renew(ts);
+            let term = self.term();
+            self.mut_leader_lease()
+                .maybe_new_remote_lease(term)
+                .map(ReadProgress::leader_lease)
+        };
+        if let Some(progress) = progress {
+            let mut meta = store_meta.lock().unwrap();
+            let reader = meta.readers.get_mut(&self.region_id()).unwrap();
+            self.maybe_update_read_progress(reader, progress);
+        }
+        if let Some(progress) = read_progress {
+            let mut meta = store_meta.lock().unwrap();
+            let reader = meta.readers.get_mut(&self.region_id()).unwrap();
+            self.maybe_update_read_progress(reader, progress);
+        }
+    }
+
+    fn maybe_update_read_progress(&self, reader: &mut ReadDelegate, progress: ReadProgress) {
+        if self.pending_remove() {
+            return;
+        }
+        debug!(
+            "update read progress";
+            "region_id" => self.region_id(),
+            "peer_id" => self.peer_id(),
+            "progress" => ?progress,
+        );
+        reader.update(progress);
+    }
+
+    fn ready_to_handle_unsafe_replica_read(&self, read_index: u64) -> bool {
+        // Wait until the follower applies all values before the read. There is still a
+        // problem if the leader applies fewer values than the follower, the follower
+        // read could get a newer value, and after that, the leader may read a stale
+        // value, which violates linearizability.
+        self.store_applied_index() >= read_index
+            // If it is in pending merge state(i.e. applied PrepareMerge), the data may be stale.
+            // TODO: Add a test to cover this case
+            && !self.has_pending_merge_state()
+            // a peer which is applying snapshot will clean up its data and ingest a snapshot file,
+            // during between the two operations a replica read could read empty data.
+            && !self.is_handling_snapshot()
```

Comment:
This is not true in v2. Because old data will be kept until all references are dropped.